dags/synapse-dataset-to-croissant.py

```diff
@@ -165,17 +199,552 @@ def set_up_logging() -> Logger:
+                otel_logger.warning(
+                    f"File {synapse_file.get('file_id')} is not anonymously accessible.")
+            else:
+                raise ex
+    return file_metadata
+
+
+def construct_distribution_section_for_files(files_attached_to_dataset: List[File], **context) -> List[Dict[str, str]]:
+    """
+    Construct the distribution section for the files attached to the dataset. This is
+    used to extract various metadata from the files and create a FileObject for each
+    file.
+
+
+    When we do not have the content_md5 for a file, we need to query Snowflake to get
+    the content_md5 for the file. This is done by querying the synapse_data_warehouse
+    database in Snowflake.
+
+    Arguments:
+        files_attached_to_dataset: The list of files attached to the dataset.
+        context: The context of the DAG run.
+
+    Returns:
+        The distribution section for the files attached to the dataset.
+    """
+    distribution_files = []
+    files_to_find_md5_in_snowflake = {}
+
+    ids_of_files = []
+    id_and_version_pairs = []
+    for file in files_attached_to_dataset:
+        file: File = file
+        if not file.file_handle or not file.file_handle.content_md5:
+            modified_file_id = int(file.id.replace("syn", ""))
+            ids_of_files.append(modified_file_id)
+            id_and_version_pairs.append(modified_file_id)
+            id_and_version_pairs.append(file.version_number)
+            files_to_find_md5_in_snowflake[int(file.id.replace(
+                "syn", ""))] = file.version_number
+
+    file_md5s = None
+    if ids_of_files:
+        snow_hook = SnowflakeHook(
+            context["params"]["snowflake_developer_service_conn"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+
+        query = f"""
+        WITH version_data AS (
+            SELECT
+                nl.id,
+                flattened.value:versionNumber::int AS versionNumber,
+                flattened.value:contentMd5::string AS contentMd5
+            FROM synapse_data_warehouse.synapse.node_latest AS nl,
+            LATERAL FLATTEN(input => nl.VERSION_HISTORY) AS flattened
+            WHERE nl.id IN ({', '.join(['%s'] * len(ids_of_files))})
+            AND nl.is_public = TRUE
+        )
+        SELECT
+            vd.id,
+            vd.versionNumber,
+            vd.contentMd5
+        FROM version_data AS vd
+        WHERE (vd.id, vd.versionNumber) IN ({', '.join(['(%s, %s)'] * (len(id_and_version_pairs)//2))});
+        """
+        try:
+            cs.execute(
+                query,
+                (ids_of_files + id_and_version_pairs),
+            )
+            file_md5s = cs.fetch_pandas_all()
+        finally:
+            cs.close()
+
+    for file in files_attached_to_dataset:
+        file: File = file
+        if file.file_handle and file.file_handle.content_md5:
+            file_md5 = file.file_handle.content_md5
```

Comment:
Is this just the case where we download the file?

Reply:
We are not downloading the file in this case, but we can get the MD5 from synapse in some cases. Those cases are when the file can be anonymously downloaded. In other cases we need to pull the data from Snowflake

Reply:
Maybe premature simplification but would it be more simple to just pull everything from snowflake instead?

(Future ticket if yes)

Reply:
Completed in https://github.com/Sage-Bionetworks-Workflows/orca-recipes/pull/84/files#r2031909088